xml_struct_derive/src/properties.rs

```diff
@@ -0,0 +1,323 @@
+            // combination of these properties is almost certainly a mistake.
+            errors.push(Error::new(
+                attr.span(),
+                "cannot declare namespaces for text content",
+            ));
+        }
+
+        if properties.ns_prefix_for_variants.is_some() && properties.should_serialize_as_text {
+            errors.push(Error::new(
+                attr.span(),
+                "cannot declare variant namespace prefix for text enum",
+            ));
+        }
+
+        // Combine and return errors if there are any. If none, we've
+        // successfully parsed the attributes and can return the appropriate
+        // props.
+        match errors.into_iter().reduce(|mut combined, err| {
+            combined.combine(err);
+
+            combined
+        }) {
+            Some(error) => Err(error),
+            None => Ok(properties),
+        }
+    }
+}
+
+#[derive(Debug)]
+/// A declaration of an XML namespace for a type with a derived `XmlSerialize`
+/// implementation.
+pub(crate) struct NamespaceDecl {
+    pub prefix: TokenStream,
+    pub name: TokenStream,
+}
+
+#[derive(Debug, Default)]
+/// Properties governing the serialization of a field in a struct or enum with a
+/// derived `XmlSerialize` implementation.
+pub(crate) struct FieldProps {
+    /// The type of XML structure which the field represents.
+    pub repr: FieldRepr,
+
+    /// Whether the field should be serialized with a "flat" representation.
+    ///
+    /// A flattened field will be serialized only as its content nodes, rather
+    /// than as an XML element containing those content nodes.
+    pub should_flatten: bool,
+
+    /// A prefix to add to this field's name when serialized as an element or
+    /// attribute.
+    pub namespace_prefix: Option<TokenStream>,
+}
+
+impl FieldProps {
+    /// Constructs a set of serialization properties for an enum or struct field
+    /// from its struct attributes.
+    pub(crate) fn try_from_attrs(
+        value: Vec<Attribute>,
+        is_named_field: bool,
+    ) -> Result<Self, Error> {
+        // Find the attribute for configuring behavior of the derivation, if
+        // any.
+        let attr = match value
+            .into_iter()
+            .find(|attr| attr.path().is_ident(MACRO_ATTRIBUTE))
+        {
+            Some(attr) => attr,
+
+            // If we don't find a matching attribute, we assume the default set
+            // of properties.
+            None => return Ok(Self::default()),
+        };
+
+        // We build a list of errors so that we can combine them later and emit
+        // them all instead of only emitting the first.
+        let mut errors = Vec::default();
+
+        let mut properties = FieldProps::default();
+        for meta in attr.parse_args_with(Punctuated::<Meta, Comma>::parse_terminated)? {
+            match meta {
+                Meta::Path(path) => {
+                    if path.is_ident("attribute") {
```

Comment:
It's not really possible to do any sort of robust type-checking in codegen, as all we have is the tokens for the struct/enum we're deriving an impl for. That said, this ends up doing the right thing: because the generated code  indicates that we want `XmlSerializeAttr`, the compiler will error appropriately, e.g.:

```
error[E0277]: the trait bound `types::common::FolderShape: xml_struct::XmlSerializeAttr` is not satisfied
  --> src/types/sync_folder_hierarchy.rs:17:23
   |
17 |     pub folder_shape: FolderShape,
   |                       ^^^^^^^^^^^ the trait `xml_struct::XmlSerializeAttr` is not implemented for `types::common::FolderShape`
   |
   = help: the following other types implement trait `xml_struct::XmlSerializeAttr`:
             i8
             i16
             i32
             i64
             u8
             u16
             u32
             u64
           and 6 others
```